File: go-backend-interview-prep/07_io_files_networking/04_paths_and_dirs/main.go
package main

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
)

func main() {
	// --- Example 1: filepath.Join ---
	fmt.Println("=== filepath.Join ===")
	p := filepath.Join("data", "users", "profile.json")
	fmt.Println("  joined:", p)
	fmt.Println("  dir:  ", filepath.Dir(p))
	fmt.Println("  base: ", filepath.Base(p))
	fmt.Println("  ext:  ", filepath.Ext(p))

	// --- Example 2: temp directory + create files ---
	fmt.Println("\n=== Temp directory ===")
	dir, err := os.MkdirTemp("", "paths-demo")
	if err != nil {
		fmt.Println("  error:", err)
		return
	}
	defer os.RemoveAll(dir) // cleanup

	// Create nested structure
	subdir := filepath.Join(dir, "sub", "nested")
	os.MkdirAll(subdir, 0755)

	// Create a few files
	for _, name := range []string{"a.txt", "b.go", "c.json"} {
		path := filepath.Join(dir, name)
		os.WriteFile(path, []byte("content of "+name), 0644)
	}
	os.WriteFile(filepath.Join(subdir, "deep.txt"), []byte("deep file"), 0644)

	fmt.Println("  created temp dir:", dir)

	// --- Example 3: WalkDir ---
	fmt.Println("\n=== WalkDir ===")
	filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		// Show relative path for cleaner output
		rel, _ := filepath.Rel(dir, path)
		if d.IsDir() {
			fmt.Printf("  [dir]  %s/\n", rel)
		} else {
			info, _ := d.Info()
			fmt.Printf("  [file] %s (%d bytes)\n", rel, info.Size())
		}
		return nil
	})

	// --- Example 4: glob pattern matching ---
	fmt.Println("\n=== Glob ===")
	matches, _ := filepath.Glob(filepath.Join(dir, "*.go"))
	for _, m := range matches {
		fmt.Println("  match:", filepath.Base(m))
	}
}

File: go-backend-interview-prep/07_io_files_networking/04_paths_and_dirs/main.py
"""Paths and directories -- Python equivalent of the Go example."""

import glob
import os
import shutil
import tempfile
from pathlib import Path


def main() -> None:
    # --- Example 1: pathlib.Path / os.path.join ---
    print("=== pathlib.Path ===")
    p = Path("data") / "users" / "profile.json"
    print(f"  joined: {p}")
    print(f"  dir:   {p.parent}")
    print(f"  base:  {p.name}")
    print(f"  ext:   {p.suffix}")

    # --- Example 2: temp directory + create files ---
    print("\n=== Temp directory ===")
    tmpdir = tempfile.mkdtemp(prefix="paths-demo")

    try:
        # Create nested structure
        subdir = Path(tmpdir) / "sub" / "nested"
        subdir.mkdir(parents=True, exist_ok=True)

        # Create a few files
        for name in ["a.txt", "b.go", "c.json"]:
            (Path(tmpdir) / name).write_text(f"content of {name}")
        (subdir / "deep.txt").write_text("deep file")

        print(f"  created temp dir: {tmpdir}")

        # --- Example 3: os.walk ---
        print("\n=== os.walk ===")
        for root, dirs, files in os.walk(tmpdir):
            rel = os.path.relpath(root, tmpdir)
            print(f"  [dir]  {rel}/")
            for fname in files:
                fpath = Path(root) / fname
                size = fpath.stat().st_size
                rel_file = os.path.relpath(fpath, tmpdir)
                print(f"  [file] {rel_file} ({size} bytes)")

        # --- Example 4: glob pattern matching ---
        print("\n=== Glob ===")
        pattern = os.path.join(tmpdir, "*.go")
        for m in glob.glob(pattern):
            print(f"  match: {Path(m).name}")
    finally:
        shutil.rmtree(tmpdir, ignore_errors=True)


if __name__ == "__main__":
    main()
